features/256-auth-gear-interaction-overview-in-next/api.md

```diff
@@ -0,0 +1,440 @@
+| `roles` | `before_roles_changed_sync(user, original_user, context)`<br/>`before_roles_changed(user, original_user, context)`<br/>`after_roles_changed_sync(user, original_user, context)`<br/>`after_roles_changed(user, original_user, context)` |
+| `enable` | `before_enable_sync(user, context)`<br/>`before_enable(user, context)`<br/>`after_enable_sync(user, context)`<br/>`after_enable(user, context)`<br/>`before_disable_sync(user, context)`<br/>`before_disable(user, context)`<br/>`after_disable_sync(user, context)`<br/>`after_disable(user, context)` |
+| `password` | `before_password_changed_sync(user, original_user, context)`<br/>`before_password_changed(user, original_user, context)`<br/>`after_password_changed_sync(user, original_user, context)`<br/>`after_password_changed(user, original_user, context)` |
+| `verify` | `before_verified_sync(user, context)`<br/>`before_verified(user, context)`<br/>`after_verified_sync(user, context)`<br/>`after_verified(user, context)`<br/>`before_unverified_sync(user, context)`<br/>`before_unverified(user, context)`<br/>`after_unverified_sync(user, context)`<br/>`after_unverified(user, context)` |
+| `metadata` | `before_metadata_modified_sync(user, original_user, context)`<br/>`before_metadata_modified(user, original_user, context)`<br/>`after_metadata_modified_sync(user, original_user, context)`<br/>`after_metadata_modified(user, original_user, context)` |
+| `user_object` | `blocking_user_sync(user)`<br>`user_sync(user)` |
+
+1. Hooks listed presented here are based on a assumption that a developer could use `content.req.path` to know the reason of certain user auth data changed.
+
+   For example, a user's password changed could be due to:
+
+   1. admin reset a user's password.
+   2. a user changes password by himself.
+   3. a user forgot password, and reset password via email form.
+
+   They are corresponding to following `context.req.path`:
+
+   1. `/reset_password`
+   2. `/change_passowrd`
+   3. `/forgot_password/reset_password`
+
+2. `user_sync` and `blocking_user_sync` are special hooks, they are triggered __AFTER__ an user object is added or there is any changes to an user object, they can be used to implement an external user profile store. It also can use as one hook to capture all changes of the user auth data or metadata.
+
+   Note that `user_sync` as an `AFTER`, `ASYNC` hook, its implementaion could be a simple after async hook like other hooks, or it can be optimized by other mechanisms (batching data update), so the `original_user` and `context` arguments are removed.
+
+   `blocking_user_sync` is an `AFTER`, `SYNC` hook. `blocking_user_sync` is especially useful when consistency is important to the application. Skygear next will rollback DB if hook failed.
+
+3. And to avoid spiral request loop, it is forbidden to send request to auth gear in hooked Cloud Function.
+
+4. `context` should contain following information:
```

Comment:
Shall we add a context that is a monotonically increasing ID (per user, or global)? Something like `context.counter`.

Another option, is we make our `context.req.id` strictly monotonically increasing. It might be useful in multiple situation in future.

The reason I'm suggesting is, how are we going to handle the following case? Let's say there are `after_enabled` and `after_disabled` called concurrently:

| Req 1 | Req 2| Remarks 
| ---- | ---- | ---- |
| enable_op |  | Req 1 called |
| after_enabled_sync() | disable_op | Req 2 called |
|  | after_disabled_sync() | |
|  | [external system update user status to disabled] | For network latency reason, req 2's hook got called first |
| [external system update user status to enabled] | | Oops

In the above situation, the external software have no way to prevent the inconsistency.

It could be avoided for the `sync_user_profile` call, if we order the network call on our end. But if other users need to handle it for rest of "action-based" hook, it seems they need a monotonic increasing ID to do so?

Correct me if I'm wrong.

Reply:
Hopefully I understand this question correctly, as I know, this issue only happens when the client SDK didn't wait for previous op finish and execute the next one.

A developer should implement

```
await enable_op();
await disable_op();
```

Instead of 

```
enable_op();
disable_op();
```

So execution order is 

| Req 1 | Req 2| Remarks 
| ---- | ---- | ---- |
| enable_op |  | Req 1 called |
| after_enabled_sync() | | enable hook invoked |
|  | disable_op | Req 2 called |
|  | after_disabled_sync() | disable hook invoked |

Reply:
@kaojohnny no, this issue happens when two requests happens concurrently in race condition. e.g. one users logged in two devices at the same time and call the ops above.